daffodil-runtime1/src/main/scala/org/apache/daffodil/layers/midBitsToEndTransformer.scala

```diff
@@ -0,0 +1,339 @@
+   * In practice, it is not expected that this transform will be used on "large" blocks,
+   * so it is not worth the effort to figure out how to implement some form of streaming.
+   */
+
+  var length: Int = -1
+  var firstBitIndex: Int = -1
+  var numBits: Int = -1
+
+  /*
+   * Note the extra padding bytes. One of these bytes is just to accomodate rounding from integer division.
+   * The other one is because, to simplify the code, a portion of the encoder may attempt to read up to 7 extra bits,
+   * then discard them. (This would happen when the last bit that gets moved is the first bit of a byte in the encoded form;
+   * the encoder will attempt to read enough bits to be able to fully replace the entire byte, then mask out the lower 7 bits)
+   */
+  def tmpByteCount = numBits / 8 + 2
+  def buffSize = length + tmpByteCount
+
+  def byteShift = numBits / 8
+  def bitShift = numBits % 8
+
+  def firstBitByteIndex = (firstBitIndex / 8)
+  def lastBitByteIndex = ((firstBitIndex + numBits) / 8)
+  def firstBitBitIndex = firstBitIndex % 8
+  def lastBitBitIndex = (firstBitIndex + numBits) % 8
+
+  private def buffToHex(buff: Array[Byte]): String = {
+    buff.map(b => String.format("%02x", Byte.box(b))).mkString(" ")
+  }
+
+  private def initialize(state: ParseOrUnparseState) {
+    length = layerLengthInBytesEv.evaluate(state).toInt
+    val argsString = layerTransformArgsEv.evaluate(state)
+      .trim()
+      .split("\\ +")
+    if (argsString.length != 2) {
+      state.SDE(s"""midBitsToEnd layer transform requires exactly 2 arguements, recieved ${argsString.length}: "${argsString.mkString(",")}" """)
+    }
+    try {
+      firstBitIndex = argsString(0).toInt
+      numBits = argsString(1).toInt
+    } catch {
+      case _: NumberFormatException => state.SDE(s"""midBitsToEnd layer transform requires Integer arguements, recieved: "${argsString.mkString(",")}"""")
+    }
+
+  }
+
+  /*
+   * Scala does not seem to support bitwise operations on bytes; and instead silently upcasts to Int.
+   * Spamming toByte is insufficient since negative numbers will get padded with 1 instead of 0
+   * For this reason, we do most of our internal logic in Int, and mask out the upper 24 bits
+   */
+
+  val bitMask: Int = 0x000000ff
+
+  private def doDecode(buff: Array[Byte]): Unit = {
+
+    /*First, move the middle bits to the end.
+     *
+     * Since the layer is byte aligned, we know we are writing them a byte aligned location
+     * We will end up moving a multiple of 8 bits. This is okay, since, after we shift,
+     * any extra bytes will be left passed the length byte, and so will be ignored
+     */
+    var dstByteIndex = length
+    var srcByteIndex = firstBitByteIndex
+    while (dstByteIndex < buffSize) {
+      val srcByte = buff(srcByteIndex) & bitMask
+      val srcByteNext = buff(srcByteIndex + 1) & bitMask
+
+      val byteUpper = (srcByte << firstBitBitIndex) & bitMask
```

Comment:
Clearly this doesn't work for different bit orders.  The direction of shift has to be different. 
Was that intentional, did we not provide code for MSBF as well as LSBF? I realize the motivating use case for this is LSBF, littleEndian, so really we only need code for that case. But the other cases ought to fail with SDE that indicate the feature is not yet implemented.

Reply:
I believe the case implemented is MSBF (as you point out, elsewhere, that seems to be the case tested).

Is the endianness actually relevent to this transform? My understanding is that endianness refers to how bits are interpereted as atomic fields, but broadly speaking the the lower indexed bytes always come first. Eg. if we have 2 16-byte fields in the following datastream:

```
0000 0001  0000 0000 
0000 0000  0000 0000
```

Endianness will determine if the first field is 1 or 256, but the first field is always represented by the first 2 bytes. Given that the transform has no knowledge of where field boundaries are, it doesn't make sense to respond to endianess.

As an aside, I believe our motivating case is actually MSBF, but it is probably best we discuss that in a more private forum.

Regardless, it is probably best I add the LSBF support now, while this shifting logic is still fresh in my mind.

Reply:
You are right that byte order doesn't matter. BitOrder does since the bits are potentially spanning a byte boundary and/or smaller than a byte.  ByteOrder is constrained by BitOrder - only certain combos are allowed. Otherwise yeah, it doesn't matter.